### ISSUE TITLE
Support custom user defined ID patterns (on model and global basis)

### ISSUE THREAD

**Issue by sunil-sadasivan** [prisma/prisma#6719]
## Problem

I prefer to scope my ids with initials of the models. This helps developers and customer support reduce any confusion by knowing exactly what model an ID corresponds to.  [Stripe uses this heavily](https://gist.github.com/fnky/76f533366f75cf75802c8052b577e2a5). 

Eg:
```
usr_<id>
post_<id>
price_<id>
```

## Suggested solution

It would be nice to allow developers to override the default ID generator (as an advanced feature). 

This may be coupled with the idea to allow the user to define the ID generation. For example, using `shortid` as the id generator.  

```
import shortid from 'shortid';
prisma.generateId = () => shortid.generate();
prisma.user.generateId = () => `u_${shortid.generate()}`;
```

## Alternatives

My preference though would be to do this as close to the schema definition as possible. So another preferable option could be to allow support for other id generators + prefixes in the schema definition. 

eg:

```
model User {
    id  @id(type: "shortid", prefix: "u_")
}
```

## Additional context

Example of stripe id prefixes: 
https://gist.github.com/fnky/76f533366f75cf75802c8052b577e2a5

**Comment by florianbress** [comment]
What's the progress on this?

**Comment by Aymericr** [comment]
Bumping this up

**Comment by codams** [comment]
Lil' bump ! We would use this heavily in our apps too

**Comment by smolak** [comment]
Yes, I would love to have support for that as well. Stripe saved me a ton of time thanks to those prefixes.

**Comment by minhdtb** [comment]
Up, we need this

**Comment by smolak** [comment]
There is a matter of prefixing things using middleware. If a bigger configuration is out of the question (like setting prefix on schema level), how about adding nanoid support at least?

**Comment by smolak** [comment]
I saw there is a similar issue present here: https://github.com/prisma/prisma/issues/12125

After tweeting this: https://twitter.com/jacek_smolak/status/1544805438344122371 I was asked to write about my use-case, so here it is.

## The problem

When I saw what the possible options for generating IDs (as a string) with Prisma are, I noticed there are `cuid()` and `uuid()`. After reading about both and discovering nanoid and learning it is better (I suggest Googling it, and of course this is not 100% true [it being better] for everyone), I wasted to make it possible to be used for Prisma's schema.

## The (mid) solution

I discovered that I could mutate the data for certain actions before it hits the DB. And in order to do that, one can use [middleware](https://www.prisma.io/docs/concepts/components/prisma-client/middleware). tl;dr - for `create` action you take the payload and overwrite the `id` property with whatever you want. Here's what I did (not exact code, but you will get the idea):

```ts
export const generateModelId: Prisma.Middleware = async (params: Prisma.MiddlewareParams, next) => {
  if (params.action === "create") {
    switch (params.model) {
      case "User":
        params.args.data.id = generateUserId();
        break;

      // ...
    }
  }

  return await next(params);
};
```

And registering it:

```ts
prisma.$use(generateModelId);
```

I went this way to have separate ID generators for separate models so that I could go with **prefixes**. Why? I can't tell you how great they are if all you see is some IDs here and there in logs or in payloads etc. and not knowing exactly where they originate from. See [my previous comment](https://github.com/prisma/prisma/issues/6719#issuecomment-1163601737).

So, the `generateUserId()` (and any other) has this implementation:

```ts
import { generateId } from "./generateId";

export const USER_ID_PREFIX = "usr_";

export const generateUserId = () => generateId(USER_ID_PREFIX);
```

And the underlying `generateId` function is this:

```ts
import { customAlphabet } from "nanoid";

// I went with custom alphabet to limit mistakes with I and l and alike (not all but some)
const ALPHABET = "abcdefghijkmnopqrstuvwxyzABCDEFGHJKLMNPQRSTUVWXYZ1234567890_";

export const generateId = (prefix = "") => `${prefix}${customAlphabet(ALPHABET)()}`;
```

## Problems with that solutions

Going with middleware I saw that:

1. Since schema.prisma has no `@default(...)` value set on `@id` type of column, I had to provide a default value in all of the code where given model was present. So e.g. if I have a `User` model and it looks like this:

```prisma
User {
  id String @id
 ...
}
```

Then if I want to e.g. create a user in some parts of the code, I need to pass `id` with a value in any form, e.g.:

```ts
prisma.user.create({
  data: {
    id: 'I can put whatever here, or leave it as an empty string, as the middleware will overwrite it but I NEED to',
    email: 'john.rambo@gmail.com',
    name: 'John'
  },
});
```

That sux.

2. Upserting data. If you want to do that, then the usage looks like this:

```ts
prisma.user.upsert({
  where: {},
  update: {},
  create: {
    id: 'same situation as before',
    email: 'john.rambo@gmail.com',
    name: 'John'
  },
});
```

And then the payload in the middleware changes, and so I need to create separate switch / case model handling. That sux again.

## A better solution

### Minimal one

The minimal one would be to have a ID generator registered / added, something like:

```ts
interface CustomIdGenerator {
  name: string;
  generateId: () => string; // or a Promise<string> as some generators out there are async
}

// Strict naming convention so that it can be used in schema
const name = 'customGenerator';
const generateId = () => { ... }

prisma.addIdGenerator({ name, generateId });
```

Usage in prisma.schema:

```prisma
User {
  id String @id @default(customGenerator())
 ...
}
```

Perhaps also adding a `setIdGenerator` to Prisma's interface - for overwriting a custom existing one, if need be. Just thinking out loud.

That would be the simplest version.

### A slightly better one

This time `model` would be passed to the generator, so that if I wanted to have different prefixes set up for different models, I could do so.

```ts
interface CustomIdGenerator {
  name: string;
  generateId: ({ model }) => string; // or a Promise<string> as some generators out there are async
}

const name = 'customGenerator';
const generateId = ({ model }) => {
  switch(model) {
    case 'User':
      return 'usr_123';
      break;
  }
}

prisma.addIdGenerator({ name, generateId });
```

Usage in prisma.schema:

```prisma
User {
  id String @id @default(customGenerator())
  ...
}
```

In both cases there's no detecting what method is using it, as the generator would just be used as any other built in.

Hope this was clear :v:

**Comment by smolak** [comment]
Yes, but I don't know if you can do it for all of the DBs supported by Prisma.

**Comment by haibh-wanosoft** [comment]
Hi @Zn4rK  Please tell me prisma.addIdGenerator({ name, generateId });
where does the addIdGenerator method get it? I use nestjs x prisma x postgresql.
Thanks you

**Comment by shtse8** [comment]
Bump it. we need this instead of terrible long id which is not useful in small project.

**Comment by codams** [comment]
Not really related to the feature request but : https://github.com/Vivalldi, any reason to thumbs down every bump on this particular issue ?

**Comment by shtse8** [comment]
We are showing support of this support why thumb me down?

**Comment by Vivalldi** [comment]
In my experience, the best practice is to react with 👍🏻 on the feature request rather than just "bumping" this in a new comment every time. All a bump comment does is notify every person watching a request. It's exhausting to watchers (and to maintainers who watch every issue). The less "bump" noise a maintainer has, the more likely they are to pay attention to an issue when the relevant discussion comes up (see `Zn4rK`'s response for example). Nothing against individuals, just trying to keep the bumps to a minimum. I'll happily stop doing so as it seems you all are in favor of these bumps.

**Comment by smolak** [comment]
👍 then :)

**Comment by cymen** [comment]
@Zn4rK Thanks for post the nanoid function. I ran into a problem with default values and I realized that is due to the ending line of the function -- it should be `VOLATILE` not `STABLE` as we do not return the same results given the same arguments for all rows in a statement. With `STABLE`, when I added the new `shortId` column to a table of existing data, I ended up with the same identical value in the `shortId` column (or a migration failure if I applied `@unique` -- I'm not using `@id` as I already have an existing id column).

So the function should be:

```plpsql
CREATE EXTENSION IF NOT EXISTS pgcrypto;

CREATE OR REPLACE FUNCTION nanoid(
    prefix text DEFAULT null,
    size int DEFAULT 21
)
RETURNS text AS $$
DECLARE
id text := '';
  i int := 0;
  urlAlphabet char(64) := 'ModuleSymbhasOwnPr-0123456789ABCDEFGHNRVfgctiUvz_KqYTJkLxpZXIjQW';
  bytes bytea := gen_random_bytes(size);
  byte int;
  pos int;
BEGIN
  WHILE i < size LOOP
    byte := get_byte(bytes, i);
    pos := (byte & 63) + 1; -- + 1 because substr starts at 1 for some reason
    id := id || substr(urlAlphabet, pos, 1);
    i = i + 1;
END LOOP;
IF prefix IS NULL THEN
    RETURN id;
ELSE
    RETURN CONCAT(LOWER(prefix), '_', id);
END IF;
END
$$ LANGUAGE PLPGSQL VOLATILE;
```

More on this in the PostgreSQL docs: https://www.postgresql.org/docs/current/xfunc-volatility.html

**Comment by cymen** [comment]
@Zn4rK I realized that line is from the repo you referenced -- I'll go make an issue/PR over there.

**Comment by AmreeshTyagi** [comment]
Any plan to provide this feature in near future?

Currently I am using cuid everywhere in my code for custom Ids, but now have to move to new project, so curious to know if it will be planned in near future or not.

**Comment by djdmbrwsk** [comment]
If you're using Postgres and just want a prefix on UUIDs you can do this:

1. Setup a model with a `@default` that uses [dbgenerated()](https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#dbgenerated)
    ```prisma
    model User {
      id    String    @id @default(dbgenerated("prefix_uuid('user_'::text)"))
      ...
    }
    ```

2. Run `npx prisma migrate dev --create-only` (just following the instructions [here](https://www.prisma.io/docs/guides/database/developing-with-prisma-migrate/enable-native-database-functions))

3. Paste this SQL at the top of the generated migration
    ```sql
    CREATE EXTENSION IF NOT EXISTS pgcrypto;
    
    CREATE OR REPLACE FUNCTION prefix_uuid(prefix text) RETURNS text AS $$
    BEGIN
        RETURN concat(prefix, gen_random_uuid());
    END;
    $$ LANGUAGE PLPGSQL VOLATILE;
    ```

4. Run `npx prisma migrate dev`

Thanks @Zn4rK and @cymen for your comments, partially based this on them 🙂

**Comment by gmwill934** [comment]
This would be great, another use case for this would be to generate `ids` with a custom prefix and sequences. 

`<prefix>_<id>`


```
transaction_000000001
transaction_000000002
transaction_000000003
transaction_000000004
...
transaction_000099823
```

**Comment by smolak** [comment]
In ~4 months this issue/feature request is going to be 2 years old ;) Any chances of picking it up?

**Comment by sarneeh** [comment]
> In ~4 months this issue/feature request is going to be 2 years old ;) Any chances of picking it up?
And it will be 10 years old if none of us will implement it 😉 2 years is enough to come to a conclusion that Prisma company does probably not see too much value in this or has other priorities. But the code is open source, so you can just grab the issue and implement it. Comments like yours and many others like `bump, I need this` won't help making this happen 😉

**Comment by smolak** [comment]
> > In ~4 months this issue/feature request is going to be 2 years old ;) Any chances of picking it up?
> 
> And it will be 10 years old if none of us will implement it 😉 2 years is enough to come to a conclusion that Prisma company does probably not see too much value in this or has other priorities. But the code is open source, so you can just grab the issue and implement it. Comments like yours and many others like `bump, I need this` won't help making this happen 😉

I will agree and disagree.

First - you're right. We should take matters into our own hands, as it is an Open Source project.
But also - if your users need it, and keep asking for it, what do you think should happen? ;)

**Comment by sarneeh** [comment]
> But also - if your users need it, and keep asking for it, what do you think should happen? ;)
If you're not paying for their product then you're not a user they will care about 😅 Also, I assume all enterprises that cooperate with Prisma have other channels for support and feature requests than GitHub Issues.

**Comment by pimeys** [comment]
If somebody needs nanoid, I just merged this: https://github.com/prisma/prisma-engines/pull/3556

**Comment by corepay** [comment]
> If you're using Postgres and just want a prefix on UUIDs you can do this:
> 
> 1. Setup a model with a `@default` that uses [dbgenerated()](https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#dbgenerated)
>    ```prisma
>    model User {
>      id    String    @id @default(dbgenerated("prefix_uuid('user_'::text)"))
>      // ...
>    }
>    ```
> 2. Run `npx prisma migrate dev --create-only` (just following the instructions [here](https://www.prisma.io/docs/guides/database/developing-with-prisma-migrate/enable-native-database-functions))
> 3. Paste this SQL at the top of the generated migration
>    ```sql
>    CREATE EXTENSION IF NOT EXISTS pgcrypto;
>    
>    CREATE OR REPLACE FUNCTION prefix_uuid(prefix text) RETURNS text AS $$
>    BEGIN
>        RETURN concat(prefix, gen_random_uuid());
>    END;
>    $$ LANGUAGE PLPGSQL VOLATILE;
>    ```
> 4. Run `npx prisma migrate dev`
> 
> Thanks @Zn4rK and @cymen for your comments, partially based this on them 🙂

FWIW I've been prefixing my PG db is'd using MD5 random `('transaction-'::text || md5((random())::text))` as column default.  Spits out a nice manageable id like `transaction-3wcvvxnfd7m3ljo9i7y13m`.

**Comment by smolak** [comment]
> If you're using Postgres and just want a prefix on UUIDs you can do this:
> 
> 1. Setup a model with a `@default` that uses [dbgenerated()](https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#dbgenerated)
>    ```prisma
>    model User {
>      id    String    @id @default(dbgenerated("prefix_uuid('user_'::text)"))
>      // ...
>    }
>    ```
> 2. Run `npx prisma migrate dev --create-only` (just following the instructions [here](https://www.prisma.io/docs/guides/database/developing-with-prisma-migrate/enable-native-database-functions))
> 3. Paste this SQL at the top of the generated migration
>    ```sql
>    CREATE EXTENSION IF NOT EXISTS pgcrypto;
>    
>    CREATE OR REPLACE FUNCTION prefix_uuid(prefix text) RETURNS text AS $$
>    BEGIN
>        RETURN concat(prefix, gen_random_uuid());
>    END;
>    $$ LANGUAGE PLPGSQL VOLATILE;
>    ```
> 4. Run `npx prisma migrate dev`
> 
> Thanks @Zn4rK and @cymen for your comments, partially based this on them 🙂

"If you're using" is not an answer. Any solution here should be DB-agnostic and provided by Prisma'a API.

**Comment by corepay** [comment]
> "If you're using" is not an answer. Any solution here should be DB-agnostic and provided by Prisma's API.

Ok then I suppose I should just tell the boss my delivery for a prefixed id in PG will be delayed until Prisma gets around to providing a versioned release supporting the requirement.

Get outta here....and a big thank you to all the other comments here that provide hacks and workarounds to help me out.

**Comment by smolak** [comment]
> > "If you're using" is not an answer. Any solution here should be DB-agnostic and provided by Prisma's API.
> 
> Ok then I suppose I should just tell the boss my delivery for a prefixed id in PG will be delayed until Prisma gets around to providing a versioned DB-agnostic release supporting the requirement.
> 
> Get outta here....and a big thank you to all the other comments here that provide hacks and workarounds to help me out.

Or you can go with a middleware solution for now. Have a look at [my comment, where I've presented it](https://github.com/prisma/prisma/issues/6719#issuecomment-1178211695). I am using it in my project, and when Prisma will support it at some point in time, I will remove it.

**Comment by AndreyMarchuk** [comment]
@smolak btw, you could keep the default value on the model to avoid issue #1 in solution: setting `id` in the create/update payload to a random value
```
id           String    @id @default(cuid())
```
middleware will provide an actual value, so default is not used

**Comment by smolak** [comment]
@AndreyMarchuk - that is a good idea. But it has some downsides too, and a couple of them:
 - ID generation will be doubled (a tiny efficiency impact)
 - if you forget to set an ID in the middleware, you will have an inconsistency on those models (cuid vs the one you want)
 
The way that I have it set up now is that I need to provide an ID on the model, and I am setting the same enum value everywhere so that if I do forget to set an ID on the model through the middleware, I will get an exception from the DB that ID is not unique.

**Comment by Melvynx** [comment]
> > If you're using Postgres and just want a prefix on UUIDs you can do this:
> > 
> > 1. Setup a model with a `@default` that uses [dbgenerated()](https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#dbgenerated)
> >    ```prisma
> >    model User {
> >      id    String    @id @default(dbgenerated("prefix_uuid('user_'::text)"))
> >      // ...
> >    }
> >    ```
> > 2. Run `npx prisma migrate dev --create-only` (just following the instructions [here](https://www.prisma.io/docs/guides/database/developing-with-prisma-migrate/enable-native-database-functions))
> > 3. Paste this SQL at the top of the generated migration
> >    ```sql
> >    CREATE EXTENSION IF NOT EXISTS pgcrypto;
> >    
> >    CREATE OR REPLACE FUNCTION prefix_uuid(prefix text) RETURNS text AS $$
> >    BEGIN
> >        RETURN concat(prefix, gen_random_uuid());
> >    END;
> >    $$ LANGUAGE PLPGSQL VOLATILE;
> >    ```
> > 4. Run `npx prisma migrate dev`
> > 
> > Thanks @Zn4rK and @cymen for your comments, partially based this on them 🙂
> 
> FWIW I've been prefixing my PG db is'd using MD5 random `('transaction-'::text || md5((random())::text))` as column default. Spits out a nice manageable id like `transaction-3wcvvxnfd7m3ljo9i7y13m`.
> 
> EDIT: Just ran a migration in turborepo using the following custom id field:
> 
> `id String @id @default(dbgenerated("concat('prefix_', md5(random()::text))"))`
> 
> will generate an id something like `prefix_x3u6ewghvjc7wazgazimk`
> 
> you can also trim the generated id to say prefix + 10 random() chars using something like this:
> 
> ` id String @id @default(dbgenerated("concat('prefix_', substring(md5(random()::text), 0, 10))"))`
> 
> will generate a shoter id something like `prefix_x3u6ewghvjc`

The trouble with this is that with each migration, Prisma wants to add this default value. Prisma seems to not detect that the change has already been made in the database.

**Comment by Hamziss** [comment]
any progress for this?

**Comment by eraoul** [comment]
Any updates? Surprised a prefix solution hasn't been implemented yet.

**Comment by South-Paw** [comment]
would be great to see this progressed

extensions proposed in #3391 have hidden gaps (nested creates being one)

**Comment by prageeth** [comment]
For anyone stumbling on this issue, I've created an extension for this: https://www.npmjs.com/package/prisma-prefixed-ids

It also supports nest writes/creates (from v1.5.0+).
